src/pymoca/tree.py

```diff
@@ -747,6 +775,494 @@ def _check_import_rules(
+        self,
+        extends: ast.ExtendsClause,
+        modification_environment: ast.ClassModification,
+        parent: ast.InstanceClass,
+        partially=True,
+    ) -> ast.InstanceExtends:
+        # 3. For each element in the extends clauses of the current element:
+        #   1. Apply steps 1 and 2 to the element, replacing the extends clause with the extends instance
+        # 4. Lookup classes of extends and ensure it is identical to lookup result from step 3
+
+        extends_class = find_name(extends.component, parent)
+
+        if extends_class is None:
+            raise ModelicaSemanticError(
+                f"Extends name {extends.component} not found in scope {parent.full_reference()}"
+            )
+        if isinstance(extends_class, ast.Symbol):
+            raise ModelicaSemanticError(
+                f"Cannot extend a Symbol: {extends.component} in {parent.full_reference()}"
+            )
+        if str(extends_class.full_reference()) == str(parent.full_reference()):
+            raise ModelicaSemanticError(
+                f"Cannot extend class '{extends_class.full_reference()}' with itself"
+            )
+        # TODO: Check with spec about what to do with extends.class_modification in case of redeclare
+        extend_mod = self._append_modifications(
+            extends.class_modification,
+            modification_environment,
+        )
+        extends_class = self._instantiate_class(
+            extends_class, extend_mod, extends_class.parent, partially
+        )
+        extends_instance = ast.InstanceExtends(
+            visibility=extends.visibility,
+            **extends_class.__dict__,
+        )
+        # TODO: Is there another way to handle `extends.class_modification`? This seems like a hack.
+        # _instantiate_partially removes from extend_mod, but not original modification_environment
+        # Reflect changes in extend_mod back into the passed-in modification env
+        modification_environment.arguments = [
+            arg for arg in modification_environment.arguments if arg in extend_mod.arguments
+        ]
+
+        if extends_instance.type in InstanceTree.BUILTIN_TYPES:
+            if len(parent.extends) > 1:
+                raise ModelicaSemanticError(
+                    "When extending a built-in class (Real, Integer, ...) "
+                    "you cannot extend other classes as well"
+                )
+
+        # TODO: Step 4 Check class lookup before and after extends
+
+        return extends_instance
+
+    def _instantiate_symbol(
+        self,
+        symbol: ast.InstanceSymbol,
+        parent: ast.InstanceClass,
+    ) -> None:
+        """Instantiate given symbol"""
+        # This is step 6 of spec 5.6.1.4 **Recursive instantiation of components**
+        # 6. Components are recursively instantiated
+
+        assert isinstance(symbol, ast.InstanceSymbol)
+        assert isinstance(parent, (ast.InstanceClass, ast.InstanceExtends))
```

Comment:
Any inherited class containing symbols would be contained in an `InstanceExtends`. Having said that I think when we add visibility to `ast.Class`, we may substitute `InstanceClass` instead of `InstanceExtends` for the extends "unnamed nodes". TBD.

Reply:
I added an `assert not isinstance(parent, ast.InstanceExtends)` after this, and ran all test cases in `parse_test.py`. It's never triggered, so I think it's safe to assume it will only be `ast.InstanceClass` (as I assumed it was with the unnamed nodes approach).

Reply:
That is a bug then! `Symbol.parent` is not getting updated when contained in `InstanceExtends`. If you look in the tree, you'll see there are `InstanceExtends` containing `InstanceSymbol`. 

The offending code is in `_instantiate_extends` and you can see it is a bit of hack:

```Modelica
        extends_class = self._instantiate_class(
            extends_class, extend_mod, extends_class.parent, partially
        )
        extends_instance = ast.InstanceExtends(
            visibility=extends.visibility,
            **extends_class.__dict__,
        )
```

There is no loop after this to update the symbol parent. I think this gets by because the `extends_instance` above is pretty much the same as the `extends_class`.

Bottom line, the "unnamed nodes" can probably use an upgrade.

Reply:
We were both right on this one. I fixed the bug per above, but still `parent` is always an `InstanceClass`. I ended up doing the "upgrade" of unnamed nodes by adding a `visibility` attribute to `Class`, using `InstanceClass` instead of `InstanceExtends`, and removing `InstanceExtends` to simplify the code.

Reply:
That sounds good to me!